### PR TITLE
Crescendo

### PR DESCRIPTION
**Pull request by austinkline**
This PR covers changes needed to upgrade HybridCustody to Cadence 1.0

Some key points:
1. Private Paths are not going to be in Cadence 1.0, so our standard path for `getCapability` that takes a `CapabilityPath` type had to be moved to take a **Capability Controller ID** instead.
2. Because there are no private paths, we now have to use storage iteration for clearing out existing capabilities when things like ownership updates or sealing change.
3. There are 4 entitlements added to the HybridCustody contract. Three of them map to a resource, and one of them is added just for publishing an account to a parent from a child.
4. ResourceDestroyed events have been added to each core resource type. There was an issue with emitting some fields we might want in the event which might need some follow-up
5. You can no longer get a public capability without <T> so retrieving a public capability had to be moved to the CapabilityFactory interface.
6. Since HybridCustody takes in a full `AuthAccount` reference currently, we have migrated all AuthAccount types to a fully entitled `&Account` reference. It does seem to need nearly all of them with the exception of the `Contract` entitlement.

The main questions coming out of this work as I see them are around Entitlements and whether the use of each entitlement gives us the access control pattern(s) we want. Personally, I had a hard time justifying splitting entitlements beyond each resource type.

### CODE REVIEW

File: contracts/HybridCustody.cdc
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:124]: Since the capability was a restricted type before, doesn't that mean they will be migrated to `&CapabilityFactory.Manager` instead of `&{CapabilityFactory.Getter}`?  So should the type here be `Manager` instead?

File: contracts/HybridCustody.cdc
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:149]: Same comment here about migrating types

File: contracts/HybridCustody.cdc
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:224]: Should this be `access(Child)`?

File: contracts/HybridCustody.cdc
Comment (joshuahannan) [inline contracts/HybridCustody.cdc:550]: Same comment about migrations here

**Review by joshuahannan — COMMENTED** [review]
Just reviewed the contracts so far. They look pretty good! I made some comments about migrating capability types, but I'm not 100% sure about the answer, so I didn't put it on every single instance that the comment would apply to in case I'm wrong. I'll flag the other ones if it turns out I am right

File: contracts/CapabilityDelegator.cdc
Comment (bjartek) [inline contracts/CapabilityDelegator.cdc:15]: should this not go away?

**Review by bjartek — CHANGES_REQUESTED** [review]

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:51]: I would say yes to this and other variables that will no longer be in use. I'll remove them today

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:124]: We should probably lean towards keeping things as open as possible unless there is a good reason not to

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:224]: So `CapabilityFilter` is intentionally public. One such use for making this a public method would be for someone to borrow a particular filter and make sure certain provider capabilities are valid or not

File: contracts/CapabilityDelegator.cdc
Comment (austinkline) [reply contracts/CapabilityDelegator.cdc:15]: Yes, all PrivatePath types will need to be deleted

File: contracts/CapabilityDelegator.cdc
Comment (austinkline) [reply contracts/CapabilityDelegator.cdc:26]: Probably should be changed. Entitlements are the main piece in this PR that I feel needs group discussion. Because of that, I went ahead and submitted what was already done so we have _something_ to go off of

File: contracts/CapabilityDelegator.cdc
Comment (austinkline) [reply contracts/CapabilityDelegator.cdc:31]: All this returns is a type, why would it need to be restricted?

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:33]: Not opposed at all. I started writing some comments out for each entitlement then realized they might just end up changing anyway which is why they're missing right now

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:1042]: This is just so that a failure message is more clear, I've been rebuilding my habits to use panic where it makes sense instead of force unwrap so that I can send a more readable failure along

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:1073]: The reason this is being done is so that no one can:

1. Make a new Account link
2. "give" ownership to someone (maybe in exchange for something)
3. once new owner thinks they have ownership, use new account link from step 1 and revoke their ownership again

So long as we permit using the interface types over concrete types, nothing would prevent someone from defining their own version of the appropriate resources to ensure things work the way they need

File: contracts/HybridCustody.cdc
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:885]:
```suggestion
```

**Review by sisyphusSmiling — COMMENTED** [review]
Great refactor so far, thank you! I was only able to look over the contracts, but left some comments. I'm also wondering if we'll need to introduce new Factory implementations for deployment covering `Vault` & `Collection` with the migration given the move from nested resources to interfaces within FT & NFT v2 standards.

**Comment by austinkline** [comment]
Quick update to this PR that I'm back from a company offsite so we're picking this up again. 

Core questions are around entitlements (expected, I'll start focusing on them soon), and on how migrations are going to work.

For migrations, my understanding right now is that **all** capabilities are going to be migrated to their conformance types. That is `Capability<&T{I}>` will be migrated to `Capability<&{I}>` in all cases. So if we want things to migrate successfully, it is the right thing to go with `Capability<&{I}>` versus `Capability<&T>`. I will operate on this assumption and then ask Bastian or someone else on the cadence team to review what they can before we finalize all of this

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:118]: Left a note on the PR itself, let's bring conversation about this topic there so we can bring all discussion about it to one place versus across comments

File: flow.json
Comment (austinkline) [reply flow.json:1]: Up Flow! I am not sure if the organization itself wants to use our npm package, I just found it much easier than getting submodules working

File: package.json
Comment (austinkline) [reply package.json:10]: Love it, will make that change

File: contracts/factories/FTAllFactory.cdc
Comment (austinkline) [reply contracts/factories/FTAllFactory.cdc:8]: The thinking is that we shouldn't return a capability which isn't the type that was requested. If we did that, attempting to cast to the one you thought was returned will panic

File: contracts/factories/NFTProviderAndCollectionFactory.cdc
Comment (austinkline) [reply contracts/factories/NFTProviderAndCollectionFactory.cdc:19]: Copying from the other comment

The thinking is that we shouldn't return a capability which isn't the type that was requested. If we did that, attempting to cast to the one you thought was returned will panic

File: contracts/factories/FTAllFactory.cdc
Comment (austinkline) [reply contracts/factories/FTAllFactory.cdc:1]: good point, yeah we can definitely add those two in as well

File: contracts/factories/FTAllFactory.cdc
Comment (austinkline) [reply contracts/factories/FTAllFactory.cdc:19]: Replied to the wrong thread earlier 🤦 

The thinking is that we shouldn't return a capability which isn't the type that was requested. If we did that, attempting to cast to the one you thought was returned will panic

File: contracts/factories/FTAllFactory.cdc
Comment (austinkline) [reply contracts/factories/FTAllFactory.cdc:8]: That bug is supposed to be fixed now, but I'll confirm the order today and fix it if things don't look right, better to keep everything consistent

File: contracts/factories/FTAllFactory.cdc
Comment (austinkline) [reply contracts/factories/FTAllFactory.cdc:1]: Added 4 new factories for the following types:

- `auth(FT.Withdraw) &{FT.Vault}`
- `&{FT.Vault}`
- `auth(NFT.Withdraw) &{Collection}`
- `&{NFT.Collection}`

File: contracts/HybridCustody.cdc
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:1070]: Nice to see this in practice! Looks much cleaner than storage iteration.

File: contracts/HybridCustody.cdc
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:1103]: I didn't realize we could use the built in `Capabilities` entitlement typically used on `Account` - maybe this would be a good entitlement for `GetterPrivate` over `Get`?

**Review by sisyphusSmiling — COMMENTED** [review]
Only got through the contracts again but can revisit tomorrow to cover txns & scripts. The entitlement assignments make sense to me, they look well done! Just have one question about `GetterPrivate` using `Get` vs `Capabilities` - curious what you think there.

File: contracts/HybridCustody.cdc
Comment (austinkline) [reply contracts/HybridCustody.cdc:1103]: Ah, yeah this is a lapse, I'll do another pass-through and make sure these are all good

File: contracts/standard/ExampleNFT.cdc
Comment (bjartek) [inline contracts/standard/ExampleNFT.cdc:67]: this is not view anymore, so if you want to you can remove it, but it is not a big deal

File: contracts/standard/ExampleToken.cdc
Comment (bjartek) [inline contracts/standard/ExampleToken.cdc:18]: FT emit standard events now, do we need this to emit its own?

Comment (bjartek) [inline contracts/standard/ExampleToken.cdc:18]:
```suggestion
```

File: contracts/standard/ExampleToken.cdc
Comment (bjartek) [inline contracts/standard/ExampleToken.cdc:23]:
```suggestion
```

File: contracts/standard/ExampleToken.cdc
Comment (bjartek) [inline contracts/standard/ExampleToken.cdc:113]:
```suggestion
```

File: contracts/standard/ExampleToken.cdc
Comment (bjartek) [inline contracts/standard/ExampleToken.cdc:129]:
```suggestion
```

File: contracts/CapabilityDelegator.cdc
Comment (bjartek) [inline contracts/CapabilityDelegator.cdc:25]: I don't Get (pun intended) why something where before all of these where restricted and now only some of them are. If that is now ok then by all means, but the semantics change.

**Review by bjartek — APPROVED** [review]

File: contracts/CapabilityDelegator.cdc
Comment (joshuahannan) [inline contracts/CapabilityDelegator.cdc:17]: I feel like this entitlement needs a slightly more descriptive name, but I'm not 100% sure of my opinion

**Review by joshuahannan — APPROVED** [review]

File: contracts/CapabilityDelegator.cdc
Comment (austinkline) [reply contracts/CapabilityDelegator.cdc:25]: I am assuming you mean

```
access(all) view fun findFirstPrivateType(_ type: Type): Type?
```

Really it's just about what something can be used while moving things to entitlements. Since this one is only returning a Type object, I don't see why we would need to make it non-public personally

File: contracts/standard/ExampleToken.cdc
Comment (austinkline) [reply contracts/standard/ExampleToken.cdc:18]: Personally I am going for the least disruptive changes required for this contract to be ready. On that basis, what a standard contract that's used just for testing does wasn't really a concern to me

I personally would rather update these at a later date since they don't have much to do with how HybridCustody itself will operate

File: contracts/CapabilityFactory.cdc
Comment (bluesign) [inline contracts/CapabilityFactory.cdc:20]: can't delete `PrivatePath` probably

Comment (austinkline) [reply contracts/CapabilityFactory.cdc:20]: I thought deleting fields was valid, but I'm guessing we will find out if this will work once I stage for migration

Comment (bluesign) [reply contracts/CapabilityFactory.cdc:20]: ah yeah you are right, I mixed

File: contracts/HybridCustody.cdc
Comment (btspoony) [inline contracts/HybridCustody.cdc:919]: I have a question: here it marks `view` function

I am a bit confused, of course, it is used for pre/post conditions.

But what I want to ask is, will it affect the normal use of this AuthAccount Reference? 
Will writing operations be unable to be performed?

Comment (sisyphusSmiling) [reply contracts/HybridCustody.cdc:919]: From my understanding, `view` restricts only the scope of the function - mutating methods can call `view` methods, but `view` methods cannot call mutating methods. So even though the reference is retrieved within this scope as a read-only return, a caller could then execute state change on the returned reference.

Comment (btspoony) [reply contracts/HybridCustody.cdc:919]: Got it, thank you!

**Review by sisyphusSmiling — APPROVED** [review]
Great work!

**Comment by sisyphusSmiling** [comment]
Update after latest migration - looks like migration [failed](https://github.com/onflow/cadence/blob/master/migrations_data/staged-contracts-report-2024-05-01T13-03-00Z-testnet.md?plain=1#L196). A couple failures stand out:

1. References to `NonFungibleToken.Owner` entitlement caused failures in `NFTProviderFactory` and `NFTProviderAndCollectionFactory`
2. Disjunction access modifiers in `CapabiltyFactory`, `CapabiltyFilter` and `CapabiltyDelegator` resulted in unsafe entitlements migration.
3. Capabilities no longer return optional values which resulted in failures within `HybridCustody`, but I'm sure there are occurrences elsewhere throughout transactions/scripts where capability retrieval is optionally chained
4. `OwnedAccount.removeParent` is `access(Owner | Remove)` while `OwnedAccountPrivate` interface modifier is simply `Owner`

I think addressing these should fix the migration issue. We'll need a new pre-release version of CLI to validate the changes via Cadence tests though since the capability retrieval pattern has been updated since the current `v1.18.0-cadence-v1.0.0-preview.22`

**Comment by austinkline** [comment]
> Update after latest migration - looks like migration [failed](https://github.com/onflow/cadence/blob/master/migrations_data/staged-contracts-report-2024-05-01T13-03-00Z-testnet.md?plain=1#L196). A couple failures stand out:
> 
> 1. References to `NonFungibleToken.Owner` entitlement caused failures in `NFTProviderFactory` and `NFTProviderAndCollectionFactory`
> 2. Disjunction access modifiers in `CapabiltyFactory`, `CapabiltyFilter` and `CapabiltyDelegator` resulted in unsafe entitlements migration.
> 3. Capabilities no longer return optional values which resulted in failures within `HybridCustody`, but I'm sure there are occurrences elsewhere throughout transactions/scripts where capability retrieval is optionally chained
> 4. `OwnedAccount.removeParent` is `access(Owner | Remove)` while `OwnedAccountPrivate` interface modifier is simply `Owner`
> 
> I think addressing these should fix the migration issue. We'll need a new pre-release version of CLI (currently in progress) to validate the changes via Cadence tests though since the capability retrieval pattern has been updated since the current `v1.18.0-cadence-v1.0.0-preview.22`

Thanks for summarizing 🙏 

I believe I've addressed everything for this week's migration, but if we're able to dry-run it locally that's probably a good idea as well

File: contracts/HybridCustody.cdc
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:532]: Revisiting this after the latest local migration result, I think this will have to be `CapabilityFactory.Manager`. I suppose that makes sense since it's at restrictive of the type as the original field type. The same will have to be the case with `delegator` - I believe that will have to be `Delegator`

**Comment by sisyphusSmiling** [comment]
Local migration passed! All contracts up to this point have been staged on Testnet, so we'll get a sanity check in the next network migration run.

**Comment by sisyphusSmiling** [comment]
Following up on office hours discussion about merging these changes. Looks like the past two migrations have been successful. Maybe we a last call for reviews before merging - thoughts @austinkline?

**Comment by austinkline** [comment]
> Following up on office hours discussion about merging these changes. Looks like the past two migrations have been successful. Maybe we a last call for reviews before merging - thoughts @austinkline?

Good with me! I'll reach out to folks and see if we can't get some approvals/comments going

**Review by bjartek — APPROVED** [review]

**Comment by austinkline** [comment]
Merging! Thanks everyone for the feedback and participation in this PR. We still have time until Crescendo goes live, so if there are any other things that come up or need addressing, please open a ticket and flag it 🙏 

https://discord.com/channels/613813861610684416/1162086721471647874/1245041945013780601

**Pull request merged**
